### PR TITLE
Groups courses users

### PR DESCRIPTION
**Pull request by James-Koval**

### CODE REVIEW

File: src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesMapper.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package ita.softserve.course_evaluation_analytics.mapper;
+
+import ita.softserve.course_evaluation_analytics.entity.GroupsCourses;
+import org.springframework.jdbc.core.RowMapper;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class GroupsCoursesMapper implements RowMapper<GroupsCourses> {
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesMapper.java]: I suggest you name it ```GroupsCoursesResultSetExtractor```

File: src/main/java/ita/softserve/course_evaluation_analytics/mapper/UsersFeedbacksMapper.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package ita.softserve.course_evaluation_analytics.mapper;
+
+
+import ita.softserve.course_evaluation_analytics.entity.UsersFeedbacks;
+import org.springframework.jdbc.core.RowMapper;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class UsersFeedbacksMapper implements RowMapper<UsersFeedbacks> {
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation_analytics/mapper/UsersFeedbacksMapper.java]: same here

File: src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+import javax.sql.DataSource;
+import java.util.List;
+
+@Repository
+public class ComplexChartRepositoryImpl implements ComplexChartRepository {
+
+    private NamedParameterJdbcTemplate namedParameterJdbcTemplate;
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java]: make in ```final```. You should always make such variables ```final```

File: src/main/java/ita/softserve/course_evaluation_analytics/service/impl/ComplexChartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class ComplexChartServiceImpl implements ComplexChartService {
+    private final ComplexChartRepository complexChartRepository;
+
+    public List<GroupsCourses> countGroupsCoursesAndUsersFeedbacks() {
+        List<GroupsCourses> groupsCoursesList = complexChartRepository.getCoursesInGroups();
+        for (GroupsCourses groupsCourses : groupsCoursesList) {
+            List<UsersFeedbacks> usersFeedbacksList = complexChartRepository.getFeedbacksFromUser();
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation_analytics/service/impl/ComplexChartServiceImpl.java]: are you going to query the database on each loop iteration? 
Bad idea 🥲

**Review by Forfend — CHANGES_REQUESTED** [review]

File: src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java (hunk 3 above)
Comment (James-Koval) [reply src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java]: But I'm changing it in the next method

File: src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesMapper.java (hunk 1 above)
Comment (James-Koval) [reply src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesMapper.java]: Done

File: src/main/java/ita/softserve/course_evaluation_analytics/mapper/UsersFeedbacksMapper.java (hunk 2 above)
Comment (James-Koval) [reply src/main/java/ita/softserve/course_evaluation_analytics/mapper/UsersFeedbacksMapper.java]: Done

File: src/main/java/ita/softserve/course_evaluation_analytics/service/impl/ComplexChartServiceImpl.java (hunk 4 above)
Comment (James-Koval) [reply src/main/java/ita/softserve/course_evaluation_analytics/service/impl/ComplexChartServiceImpl.java]: Done

File: src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java (hunk 3 above)
Comment (Forfend) [reply src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java]: ```NamedParameterJdbcTemplate``` should be defined as a Bean in your configuration, and here you need just to inject, you are not supposed to define your config here

**Review by Forfend — CHANGES_REQUESTED** [review]

File: src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java (hunk 3 above)
Comment (James-Koval) [reply src/main/java/ita/softserve/course_evaluation_analytics/repository/impl/ComplexChartRepositoryImpl.java]: Got it ))

File: src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesResultSetExtractor.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import org.springframework.jdbc.core.RowMapper;
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+public class GroupsCoursesResultSetExtractor implements RowMapper<GroupsCourses> {
+    @Override
+    public GroupsCourses mapRow(ResultSet resultSet, int i) throws SQLException {
+        return new GroupsCourses(resultSet.getLong("id"),
+                resultSet.getString("group_name"),
```
Comment (Forfend) [inline src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesResultSetExtractor.java]: It's better to extract these magic names into constants

**Review by Forfend — CHANGES_REQUESTED** [review]

File: src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesResultSetExtractor.java (hunk 5 above)
Comment (James-Koval) [reply src/main/java/ita/softserve/course_evaluation_analytics/mapper/GroupsCoursesResultSetExtractor.java]: Done

**Review by Forfend — APPROVED** [review]

**Pull request merged**